### DISCUSSION TITLE
Updating defaultValues does not react.

### DISCUSSION THREAD

**Discussion by kobe651jp** [react-hook-form/react-hook-form#9294] [Q&A] [answered]
When updating a resDataElement set to defaultValues, getValues() will only return the previous values.

I think there is something wrong with the settings. I checked past posts, but could not find a solution. 

Here is my code:

```
  const {
    register,
    getValues,
    setValue,
    control,
    reset,
    formState: { errors, isDirty, isValid },
  } = useForm({
    defaultValues:{
      ...resDataElement
    },
    mode: 'onChange',
    reValidateMode: 'onChange',
    resolver: yupResolver(ElmSchema),
  });
```

**Comment by dorilahav** [comment, marked as answer]
Hey @kobe651jp 
Default values are meant to be the values that are used before any other value is provided. Because of that, in order to prevent overriding user input, default values are not reactive on purpose.
If you'd want to fill the form with default values every time they change, you could call the `reset` function provided by `useForm`.
The `reset` function can take a defaultValues argument and set the form values accordingly.
To get your desired behevior I would suggest using a useEffect with resDataElement as a dependency that is calling said `reset` function. Example:
```js
useEffect(() => {
  form.reset({...resDataElement});
}, [resDataElement]);
```

**Reply by bluebill1049** [reply]
Thanks, @dorilahav that's correct! This maybe interested to you guys: https://github.com/react-hook-form/react-hook-form/pull/9261

**Reply by bennettdams** [reply]
Is it correct that the remote data (which represents the "new" default values after submit) has to be memoized? I encountered an infinite `useEffect` without it. This was because my default values are based on an array. Just wondering if there's a better way/if I'm doing something wrong?

**Comment by kobe651jp** [comment]
Hi,@dorilahav 
As per your advice, I've seen using "reset" to handle resDataElement changes. It works fine when {} is changed to {...}. 
However, when {...} is changed to {}, it becomes an infinite loop.

I tried using the "values" props, but it doesn't work as expected.

The point is that when resDataElement becomes {} empty, it becomes an infinite loop.

Here is my code 
```
  const {
    register,
    getValues,
    reset,
    formState: { errors, isDirty, isValid },
  } = useForm({
    defaultValues:{
      ...resDataElement
    },
    // values:{
    //   ...resDataElement
    // },
    mode: 'onChange',
    reValidateMode: 'onChange',
    resolver: yupResolver(ElmSchema),
  });

  // defaultValuesは初期値のみ有効なため。
  useEffect(()=>{
    reset({
      ...resDataElement
    });
  },[resDataElement,reset]);

  console.log('resDataElement', resDataElement);
```
#9261 method seems helpful if it's works normally.

**Reply by dorilahav** [reply]
I tried copying most of your code on code sandbox and I can't seem to replicate the infinite-loop you are referring to.
Can you please include more of your code? It is unclear to me how it happens.
https://codesandbox.io/s/mystifying-ully-s2070k

**Comment by kobe651jp** [comment]
I found a solution.

from 
```
  const { data: resDataElement = {} } = useQuery(
    ['element', resDataNav.xnav_link?.xelm_uuid],
    async () => {
      const resData = await api.postElements({
        data: {
          xelm_uuid: resDataNav.xnav_link?.xelm_uuid,
        },
        method: 'read',
      });
      return resData;
    },
    {
      enabled: !!resDataNav.xnav_link?.xelm_uuid,
      retry: 1,
      retryDelay: 1000, // Will always wait 1000ms to retry, regardless of how many retries
      cacheTime: 24 * 3600 * 1000,
      staleTime: 24 * 3600 * 1000,
    },
  );
```
to
```
  const memorizedEmpty = useMemo(()=>({}),[]);
  const { data: resDataElement = memorizedEmpty } = useQuery(
    ['element', resDataNav.xnav_link?.xelm_uuid],
    async () => {
      const resData = await api.postElements({
        data: {
          xelm_uuid: resDataNav.xnav_link?.xelm_uuid,
        },
        method: 'read',
      });
      return resData;
    },
    {
      enabled: !!resDataNav.xnav_link?.xelm_uuid,
      retry: 1,
      retryDelay: 1000, // Will always wait 1000ms to retry, regardless of how many retries
      cacheTime: 24 * 3600 * 1000,
      staleTime: 24 * 3600 * 1000,
    },
  );
```
My guess is if not use memorized empty value(resDataElement), in follows code resDataElement change every render(for useEffect).
```
  useEffect(()=>{
    reset({...resDataElement});
  },[resDataElement]);

```
Many thanks.
